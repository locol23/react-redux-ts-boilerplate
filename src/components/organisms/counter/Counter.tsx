import * as React from 'react'
import { lifecycle } from 'recompose'
import styled from 'styled-components'
import ButtonsContainer from '../../molcules/ButtonsContainer'

interface Props {
  title: string
  value: number
  setTitle(s: string): void
  increment(): void
  decrement(): void
  multiply(n: number): void
  reset(): void
}

const Component: React.SFC<Props> = props => {
  return (
    <Container>
      <Title>{props.title}</Title>
      <Value>value: {props.value}</Value>
      <ButtonsContainer {...props} />
    </Container>
  )
}

export const Counter: React.ComponentType<Props> = lifecycle<Props, {}, {}>({
  componentDidMount() {
    this.props.setTitle('Counter')
  },
})(Component)

const Container = styled.div`
  padding: 10px;
`

const Title = styled.h1`
  color: #333333;
`

const Value = styled.div`
  text-align: center;
  font-size: 25px;
`
