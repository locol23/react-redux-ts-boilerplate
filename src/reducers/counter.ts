// Action
const INCREMENT = 'counter/increment'
const DECREMENT = 'counter/decrement'
const MULTIPLY = 'counter/multiply'
const RESET = 'counter/reset'

// Types
interface Increment {
  type: typeof INCREMENT
}

interface Decrement {
  type: typeof DECREMENT
}

interface Multiply {
  type: typeof MULTIPLY
  payload: {
    amount: number
  }
}

interface Reset {
  type: typeof RESET
}

type Action = Increment | Decrement | Multiply | Reset

// Action Creator
export const increment = (): Increment => ({ type: INCREMENT })

export const decrement = (): Decrement => ({ type: DECREMENT })

export const multiply = (amount: number): Multiply => ({
  type: MULTIPLY,
  payload: {
    amount,
  },
})

export const reset = (): Reset => ({ type: RESET })

// State
export interface State {
  value: number
}

const initialState: State = {
  value: 0,
}

export default (state: State = initialState, action: Action) => {
  switch (action.type) {
    case INCREMENT: {
      return {
        ...state,
        value: state.value + 1,
      }
    }
    case DECREMENT: {
      return {
        ...state,
        value: state.value - 1,
      }
    }
    case MULTIPLY: {
      return {
        ...state,
        value: state.value * action.payload.amount,
      }
    }
    case RESET: {
      return {
        ...initialState,
      }
    }
    default: {
      return state
    }
  }
}
