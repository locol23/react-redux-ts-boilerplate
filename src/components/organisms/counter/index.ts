import { connect } from 'react-redux'
import { State as RootState } from '../../../reducers'
import * as CounterActions from '../../../reducers/counter'
import Counter from './Counter'

interface OuterProps {}

interface Props {
  value: number
}

// connect
const mapStateToProps = (state: RootState, props: OuterProps): Props => {
  return {
    value: state.counter.value,
  }
}

const Enhancer = connect(
  mapStateToProps,
  {
    increment: CounterActions.increment,
    decrement: CounterActions.decrement,
    multiply: CounterActions.multiply,
    reset: CounterActions.reset,
  }
)

export default Enhancer(Counter)
