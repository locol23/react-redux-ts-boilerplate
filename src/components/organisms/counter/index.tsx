import * as React from 'react'
import { AllAction, connector, RootState } from '../../../reducers'
import { Counter as CounterContent } from './Counter'

const mapStateToProps = (state: RootState) => {
  return {
    title: state.setting.title,
    value: state.counter.value,
  }
}

const selectAction = (actions: AllAction) => {
  return {
    setTitle: actions.setting.setTitle,
    increment: actions.counter.increment,
    decrement: actions.counter.decrement,
    multiply: actions.counter.multiply,
    reset: actions.counter.reset,
  }
}

export const Counter = connector(mapStateToProps, selectAction)(
  (props: any) => <CounterContent {...props} />
)
