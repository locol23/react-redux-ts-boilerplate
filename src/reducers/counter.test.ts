import { decrement, increment, multiply, reducer, reset } from './counter'

test('should return the initial state', () => {
  expect(reducer(undefined, { type: undefined as any })).toMatchSnapshot()
})

test('should handle INCREMENT', () => {
  expect(reducer(undefined, increment())).toMatchSnapshot()
})

test('should handle DECREMENT', () => {
  expect(reducer(undefined, decrement())).toMatchSnapshot()
})

test('should handle MULTIPLY', () => {
  expect(reducer(undefined, multiply(2))).toMatchSnapshot()
})

test('should handle RESET', () => {
  expect(reducer(undefined, reset())).toMatchSnapshot()
})
