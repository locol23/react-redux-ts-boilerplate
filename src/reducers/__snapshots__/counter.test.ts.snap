// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should handle DECREMENT 1`] = `
Object {
  "value": -1,
}
`;

exports[`should handle INCREMENT 1`] = `
Object {
  "value": 1,
}
`;

exports[`should handle MULTIPLY 1`] = `
Object {
  "value": 0,
}
`;

exports[`should handle RESET 1`] = `
Object {
  "value": 0,
}
`;

exports[`should return the initial state 1`] = `
Object {
  "value": 0,
}
`;
